### PR TITLE
Add Additional Tests to UprobesUnwindingVisitor

### PR DESCRIPTION
**Pull request by florian-kuebler**
Test: Run tests
Bug: http://b/185088047

### CODE REVIEW

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 1]
```diff
@@ -127,31 +162,15 @@ TEST_F(UprobesUnwindingVisitorTest, VisitValidStackSampleWithoutUprobesEmitsEven
   };
   event.ring_buffer_record->sample_id = sample_id;
 
+  EXPECT_CALL(return_address_manager_, PatchSample).WillRepeatedly(Return());
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:165]: What is this for? Is this actually required for behavior or is it for the "uninteresting calls" warnings?

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 2]
```diff
@@ -240,20 +260,54 @@ TEST_F(UprobesUnwindingVisitorTest,
...
+  EXPECT_EQ(unwinding_errors, 1);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 0);
+}
+
+TEST_F(UprobesUnwindingVisitorTest, VisitStackSampleWithinUprobeLeadsToUnwindingError) {
+  constexpr uint32_t kPid = 10;
+  constexpr uint64_t kStackSize = 13;
+  StackSamplePerfEvent event{kStackSize};
+  perf_event_sample_id_tid_time_streamid_cpu sample_id{
+      .pid = kPid, .tid = 11, .time = 15, .stream_id = 12, .cpu = 0, .res = 0};
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:292]: Add trailing comma here and in the previous test.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 3]
```diff
@@ -265,8 +319,302 @@ TEST_F(UprobesUnwindingVisitorTest,
...
+  visitor_->visit(&event);
+
+  EXPECT_THAT(actual_callstack_sample.callstack().pcs(),
+              ElementsAre(kTargetAddress1, kTargetAddress2, kTargetAddress3));
+
+  EXPECT_EQ(unwinding_errors, 0);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 0);
+}
+
+TEST_F(UprobesUnwindingVisitorTest, VisitBrokenCallchainSampleDiscardsEvents) {
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:376]: I would replace "broken" with an additional detail, i.e., that it's a single-frame callchain. But see the other comment: have we observed this?

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 4]
```diff
@@ -265,8 +319,302 @@ TEST_F(UprobesUnwindingVisitorTest,
...
+  visitor_->visit(&event);
+
+  EXPECT_THAT(actual_callstack_sample.callstack().pcs(),
+              ElementsAre(kTargetAddress1, kTargetAddress2, kTargetAddress3));
+
+  EXPECT_EQ(unwinding_errors, 0);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 0);
+}
+
+TEST_F(UprobesUnwindingVisitorTest, VisitCallchainSampleWithBrokenUprobeDiscardEvents) {
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:477]:
```suggestion
TEST_F(UprobesUnwindingVisitorTest, VisitCallchainSampleWithBrokenUprobeDiscardsEvents) {
```

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 5]
```diff
@@ -265,8 +319,302 @@ TEST_F(UprobesUnwindingVisitorTest,
...
+  visitor_->SetUnwindErrorsAndDiscardedSamplesCounters(&unwinding_errors,
+                                                       &discarded_samples_in_uretprobes_counter);
+
+  visitor_->visit(&event);
+
+  EXPECT_EQ(unwinding_errors, 1);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 0);
+}
+
+TEST_F(UprobesUnwindingVisitorTest, VisitCallchainSampleInsideUprobeCodeDiscardEvents) {
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:523]:
```suggestion
TEST_F(UprobesUnwindingVisitorTest, VisitCallchainSampleInsideUprobeCodeDiscardsEvents) {
```

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp [hunk 6]
```diff
@@ -265,8 +319,302 @@ TEST_F(UprobesUnwindingVisitorTest,
...
+                                                       &discarded_samples_in_uretprobes_counter);
+
+  visitor_->visit(&event);
+
+  EXPECT_EQ(unwinding_errors, 0);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 1);
+}
+
+TEST_F(UprobesUnwindingVisitorTest,
+       DISABLED_VisitLeafCallOptimizedCallchainSampleWithoutUprobesEmitsEvents) {
```
Comment (dpallotti) [inline src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:571]: Add a comment why this is disabled, possibly with a TODO linking a bug. Or just consider adding this in another change.

File: src/LinuxTracing/UprobesReturnAddressManager.h [hunk 7]
```diff
@@ -67,8 +68,13 @@ class UprobesReturnAddressManager {
   // This function patches the callchain, using the maps information to identify
   // instruction pointers of uprobe code and using the return address saved in
   // the uprobes.
-  bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
-                      unwindstack::Maps* maps) {
+  virtual bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
+                              orbit_linux_tracing::LibunwindstackMaps* maps) {
+    if (callchain_size == 0) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesReturnAddressManager.h:73]: It should not happen, but I add the CHECK for callchain not being nullptr, and would argue (nullptr, 0) could be reasonable arguments.

File: src/LinuxTracing/UprobesReturnAddressManager.h [hunk 8]
```diff
@@ -67,8 +68,13 @@ class UprobesReturnAddressManager {
...
+      return true;
+    }
+    CHECK(callchain != nullptr);
+    CHECK(maps != nullptr);
     std::vector<uint64_t> frames_to_patch;
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesReturnAddressManager.h:78]: Done.

[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesReturnAddressManager.h]: Done.

File: src/LinuxTracing/UprobesUnwindingVisitor.cpp [hunk 9]
```diff
@@ -102,13 +102,19 @@ void UprobesUnwindingVisitor::visit(CallchainSamplePerfEvent* event) {
...
+  if (!return_address_manager_->PatchCallchain(event->GetTid(), event->GetCallchain(),
+                                               event->GetCallchainSize(), current_maps_)) {
+    if (unwind_error_counter_ != nullptr) {
+      ++(*unwind_error_counter_);
+    }
     return;
   }
 
   // The top of a callchain is always inside the kernel code.
   if (event->GetCallchainSize() <= 1) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: IIRC, this is something we saw. However, I agree that we should log this. But, I guess we still want to discard the sample here?

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 1 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:165]: That is for the warning only.

File: src/LinuxTracing/UprobesReturnAddressManager.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/UprobesReturnAddressManager.h:73]: With `CHECK(callchain != nullptr);` you're saying that they're not reasonable arguments. If you have `CHECK(callchain != nullptr);`, why not also have `CHECK(callchain_size > 0);` if the two parameters go together?

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 2 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp]: Done.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp]: Renamed it. IIRC, we've seen this, yes.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 4 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp]: Done.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 5 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp]: Done.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 6 above)
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:584]: Done.

File: src/LinuxTracing/UprobesReturnAddressManager.h [hunk 10]
```diff
@@ -67,8 +68,13 @@ class UprobesReturnAddressManager {
   // This function patches the callchain, using the maps information to identify
   // instruction pointers of uprobe code and using the return address saved in
   // the uprobes.
-  bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
-                      unwindstack::Maps* maps) {
+  virtual bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
+                              orbit_linux_tracing::LibunwindstackMaps* maps) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesReturnAddressManager.h]: Not if I add the namespace around the class, what I did now. Thanks.

File: src/LinuxTracing/UprobesReturnAddressManager.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesReturnAddressManager.h:74]: Fine for me, added the additional check.

File: src/LinuxTracing/UprobesUnwindingVisitor.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: Yes, we still want to discard it. But I'm not sure whether it should add to the counter as I'm not sure what an empty callchain means.
I feel like we should try to have a better idea. For example, is it frequent? If yes, the `ERROR` you added would now spam the logs.

File: src/LinuxTracing/UprobesUnwindingVisitorTest.cpp (hunk 1 above)
Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:165]: I think that `.Times(AnyNumber())` is clearer then (if it works, haven't tried it).

Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:165]: Or just .Times(1)? as I expect only one call?

Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitorTest.cpp:165]: Probably even better then.

File: src/LinuxTracing/UprobesUnwindingVisitor.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: So I tried in our default benchmark and did not hit a single time this line in a 2min capture. I would still like to keep this error.

[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: If this is rare or unexpected I agree we should keep the `ERROR` as suggested above, but we need to make sure that's the case. Could you try on Trata or something similar with frame pointers (even if it doesn't have frame pointers, it doesn't matter)?

[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: >Could you try on Trata or something similar with frame pointers (even if it doesn't have frame pointers, it doesn't matter)?

That's what I meant with our default benchmark.

[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: Aaah, perfect then.

**Review by dpallotti — APPROVED** [review]

File: src/LinuxTracing/UprobesUnwindingVisitor.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:114]: Maybe add a comment that we don't expect that.

[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/LinuxTracing/UprobesUnwindingVisitor.cpp:115]: Added a comment.

**Pull request merged**